platform/engineering/design-docs/2021-06-11-small-batch-publishing.md

```diff
@@ -0,0 +1,350 @@
+
+
+Editable link: [https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#](https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#)
+
+
+## **Specifics**
+
+
+### **Detailed Design**
+
+
+#### Drupal Triggers Content Publishing
+
+When an editor performs an action in Drupal (ie publishing a node, edits a menu item), Drupal will determine which pages need to be published. [The logic will be encapsulated by a Drupal Service.](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5509) The rules around the dependency graph are being tracked on [here](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/dependency_graph) and are currently a work in progress. Which rules are addressed will be determined by the content type being worked on to support Small Batch Publishing.  Drupal will also have logic to not allow a Small Batch Publishing if a full build is in progress.
+
+Drupal will track the queue progress using the [advancedqueue](https://www.drupal.org/project/advancedqueue) and the Public Batch API described below. The advancedqueue module allows for observability and reporting.
+
+
+#### Preview Server Changes
+
+The code from the current preview server will be abstracted and reused in the lambda function.  The following changes will occur:
+
+*   Support building pages for a specific node revision.  This allows the lambda function to generate HTML for a published revision or any draft revision.  
+*   Add the ability to toggle [debug information](https://monosnap.com/file/e7gAeGAg7m0crd8anzjMt5Z3NiuGvL) 
+*   Code will be updated to use the json processing from here: `src/site/stages/build/drupal/*.js`
+*   `/diff` endpoint code will be refactored to use the same code as the lambda function.
+*   GraphQL Query building will be refactored to have Full and Small Batch use the same queries.
+*   The Get Drupal Content plugin should be broken apart to make the layers reusable with the different services.
+
+
+#### Facility Banner Alerts
+
+The first type of data to be updated using the Small Batch Publishing will be the facility banner alerts and alerts. These types of content currently trigger a deploy from Drupal to the jenkins job. The trigger from Drupal to Jenkins will be replaced with Small Batch Publishing.  By removing the need for Drupal to trigger the Jenkins job, we eliminate logic around Drupal bulk publish events and when an editor is allowed to trigger a Small Batch Publish.
+
+1. Banner Alert or Alert updated in Drupal
+2. Drupal sends Publish Batch Record to SQS
+3. Lambda picks up Publish Batch record
+4. Metalsmith build runs for alerts and generates a JSON file instead of an HTML file.
+5. JSON file is pushed to S3
+6. A React Widget looks for the JSON file on S3 to populate the alert.
```

Comment:
If this is a react widget on the page do we need metalsmith to run at all? Can we just update an endpoint and hit it live?

Reply:
The idea would that the react widget would load the data directly from a JSON file on S3 so no metasmith build required.  The JSON file would be updated as part of the SBP pipeline.

Reply:
Gotcha, ok, #3 confused me there

Reply:
I separated out the creation of the react widget and linked to the epic with more detail.